keps/0031-namespace-management.md

```diff
@@ -0,0 +1,156 @@
+* Define if an operator can leverage more than one namespace and under what conditions
+
+
+### Non-Goals
+
+
+## Namespace Management in the Eco-System
+
+### Kubernetes Objects and Namespaces
+
+When working kubernetes objects, be it Pods, ReplicaSets, Deployments and the like that require a namespace, the responsbility falls on the user to create that namespace prior assigning the resource or it will fail.  Available is an [auto-provision admission webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#namespaceautoprovision) available to auto-create namespaces. Users desiring auto-provisioning of namespaces should install this tool.  This webhook has the limitation of creating namespaces without extra metadata which defeats the value for use cases which need that metadata.
+
+### Helm
+
+Helm v1 and v2 provided an auto-provision capability.  The new Helm v3 (based on previous experiences) removed this capability and now requires users to create namespaces prior to installing a helm chart. Much detail is written in comments for [helm issue 6794](https://github.com/helm/helm/issues/6794)
+
+### The Many Meanings of a Namespace
+
+Challenages to auto-provision of namespaces are complicated by the many reasons / values for creating them which includes:
+
+1. Logical separation of work (separating 1000s of pods into 10s of buckets for management)
+1. Quota management
+1. Security separation
+1. Metadata (where labels and annotations provide meaning and value to controllers managing resources in a namespace)
+
+Based on the last 3 reasons, it is common for an administrator to create namespaces to manage the other concerns associated with a namespace.  It would be challeging to manage a cluster where dozen or more namespaces are adhoc being created (or deleted) which may require a host of other configurations around quota and security.
+
+## Proposals
+
+Prior to making KEP `provisional` we will chose a namespace creation proposal to move forward with and move the 2nd proposal to alternative.  For now, we have 2 competing namespace creation proposals.  
+In addition there is a proposal for the use of multiple namespaces.
+
+### Proposal 1: No namespace creation
+
+All namespace creation is managed outside KUDO. Users such as D2iQ would need to find or build automation for the creation of their namespaces. This REQUIRES that KUDO honor the deployment of resources to the "default" namespace, or to the namespace defined by `--namespace` during install.  
+
+KUDO should further require that NO namespaces be allow in manifests for resources deployed during "install".  The concept of defining manifest which land in multiple namespaces is defined in a latter proposal and is a separate concern. Without multi-namespace support, all resources are install to one namespace and namespace must be created prior to installation and that namespace must be specified during installation.
+
+### Proposal 2: Namespace Creation
```

Comment:
@kensipe do you think we can add `ClusterInstance` https://github.com/kudobuilder/kudo/issues/1493#issuecomment-626326419 as a proposal?

Reply:
yeah... I added it as an alt idea... perhaps it such be promoted to a full proposal.  I don't know if I know enough details around what you have in mind to write it up... would you like to push more detail for it?

Reply:
yeah it's good as an alternative and I can push more details if we decide to push for adding it as a proposal after tomorrow sync